src/expr.cc

```diff
@@ -463,45 +463,127 @@ void ExpressionMember::print(std::ostream &stream) const
+ExpressionFunctionCall::ExpressionFunctionCall(const std::string &name, const AssignmentList &arglist, Expression *expr)
+	: Expression(expr), funcname(name), call_arguments(arglist)
 {
 }
 
-ValuePtr ExpressionFunctionCall::evaluate(const Context *context) const
+ExpressionFunctionCall * ExpressionFunctionCall::create(const std::string &name, const AssignmentList &arglist, Expression *expr)
+{
+	if (name == "echo") {
+		return new ExpressionEcho(name, arglist, expr);
+	} else if (name == "assert") {
+		return new ExpressionAssert(name, arglist, expr);
+	} else if (name == "let") {
+		return new ExpressionLet(name, arglist, expr);
+	} else {
+		if (expr == 0) {
+			return new ExpressionSimpleFunctionCall(name, arglist);
+		} else {
+			return new ExpressionError(name, arglist, expr);
+		}
+	}
+}
+
+void ExpressionFunctionCall::print(std::ostream &stream) const
+{
+	stream << this->funcname << "(" << this->call_arguments << ")";
+	if (this->first) {
+		stream << " " << *this->first;
+	}
+}
+
+ExpressionSimpleFunctionCall::ExpressionSimpleFunctionCall(const std::string &name, const AssignmentList &arglist)
+	: ExpressionFunctionCall(name, arglist, NULL)
+{
+}
+
+ValuePtr ExpressionSimpleFunctionCall::evaluate(const Context *context) const
 {
 	if (StackCheck::inst()->check()) {
 		throw RecursionException::create("function", funcname);
 	}
-    
+
 	EvalContext c(context, this->call_arguments);
 	ValuePtr result = context->evaluate_function(this->funcname, &c);
 
 	return result;
 }
 
-void ExpressionFunctionCall::print(std::ostream &stream) const
+ExpressionError::ExpressionError(const std::string &name, const AssignmentList &arglist, Expression *expr)
+	: ExpressionFunctionCall(name, arglist, expr)
 {
-	stream << this->funcname << "(" << this->call_arguments << ")";
 }
 
-ExpressionLet::ExpressionLet(const AssignmentList &arglist, Expression *expr)
-	: Expression(expr), call_arguments(arglist)
+ValuePtr ExpressionError::evaluate(const Context * /*context*/) const
 {
+	throw RecursionException::create("error", funcname);
 }
 
-ValuePtr ExpressionLet::evaluate(const Context *context) const
+ExpressionEcho::ExpressionEcho(const std::string &name, const AssignmentList &arglist, Expression *expr)
+	: ExpressionFunctionCall(name, arglist, expr)
 {
+}
+
+ValuePtr ExpressionEcho::evaluate(const Context *context) const
+{
+	ExperimentalFeatureException::check(Feature::ExperimentalEchoExpression);
+
+	EvalContext assignment_context(context, this->call_arguments);
+
 	Context c(context);
-	evaluate_sequential_assignment(this->call_arguments, &c);
+	assignment_context.assignTo(c);
 
-	return this->first->evaluate(&c);
+	ValuePtr result = this->first ? this->first->evaluate(&c) : ValuePtr::undefined;
+
+	std::stringstream msg;
+	EvalContext echo_context(&c, this->call_arguments);
+	msg << "ECHO: " << echo_context;
+
+	if (this->first) {
+		if (echo_context.numArgs()) msg << ", ";
+		if (result->type() == Value::STRING) {
```

Comment:
Not as-is if we want the same output as the statement `echo()`. The `operator<<` uses `QuotedString` which converts `TAB` to `\t` and such.